openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -408,34 +417,66 @@ def _verify_user(request, response, userid_in_session):
-            # is logging out.  We won't log that.
-            if request.safe_cookie_verified_user_id != userid_in_session and userid_in_session is not None:
-                log.warning(
-                    (
-                        "SafeCookieData user at request '{}' does not match user in session: '{}' "
-                        "for request path '{}'"
-                    ).format(  # pylint: disable=logging-format-interpolation
-                        request.safe_cookie_verified_user_id, userid_in_session, request.path,
-                    ),
-                )
-                set_custom_attribute("safe_sessions.user_mismatch", "request-session-mismatch")
+
+            session_at_response = 'None'
+            if hasattr(request, 'session') and hasattr(request.session, 'session_key'):
+                session_at_response = request.session.session_key
+
+            if is_request_response_mismatch or is_request_session_mismatch:
+                if not cache.touch(USER_MISMATCH_CACHE_KEY, 300):
+                    cache.set(USER_MISMATCH_CACHE_KEY, True, 300)
+                session_has_changed = not request.safe_cookie_session_id == session_at_response
+                # remove the stored old session id for security
+                del request.safe_cookie_session_id
+
+                # for security, don't log the actual session id, just whether or not it changed
+                log_session_suffix = 'Session id has changed.' if session_has_changed else 'Session id has not changed'
+
+                if not is_request_session_mismatch:
+                    log.warning(
+                        (
+                            "SafeCookieData user at request '{}' does not match user in response: '{}' "
+                            "for request path '{}'. {}"
+                        ).format(  # pylint: disable=logging-format-interpolation
+                            request.safe_cookie_verified_user_id, request.user.id, request.path,
+                            log_session_suffix
+                        ),
+                    )
+                    set_custom_attribute("safe_sessions.user_mismatch", "request-response-mismatch")
+                if not is_request_response_mismatch:
+                    log.warning(
+                        (
+                            "SafeCookieData user at request '{}' does not match user in session: '{}' "
+                            "for request path '{}'. {}"
+                        ).format(  # pylint: disable=logging-format-interpolation
+                            request.safe_cookie_verified_user_id, userid_in_session, request.path,
+                            log_session_suffix
+                        ),
+                    )
+                    set_custom_attribute("safe_sessions.user_mismatch", "request-session-mismatch")
+                else:
+                    # both session user and user in response are different than user in request
```

Comment:
I believe this will actually end up running for both the "both" and the "neither" case, but I'm having some trouble with the negated logic in the first two guards.